src/main/kotlin/org/rust/lang/core/type/TypeInference.kt

```diff
@@ -1,32 +1,47 @@
 package org.rust.lang.core.type
 
-import org.rust.lang.core.psi.*
-import org.rust.lang.core.psi.util.parentOfType
+import com.intellij.psi.util.CachedValueProvider
+import com.intellij.psi.util.PsiModificationTracker
+import org.rust.lang.core.psi.RustExprElement
+import org.rust.lang.core.psi.RustTypeElement
+import org.rust.lang.core.psi.visitors.RustTypificationVisitor
+import org.rust.lang.core.type.unresolved.RustUnresolvedType
+import org.rust.lang.core.type.visitors.RustTypeResolvingVisitor
 import org.rust.lang.utils.psiCached
 
-val RustExprElement.inferredType: RustResolvedType by psiCached {
-    when (this) {
-        is RustPathExprElement -> {
-            val target = path.reference.resolve()
-            when (target) {
-                is RustSelfArgumentElement -> target.parentOfType<RustImplItemElement>()?.type?.resolvedType ?: RustUnknownType
-                else -> RustUnknownType
-            }
-        }
-        else -> RustUnknownType
+val RustExprElement.type: RustUnresolvedType by psiCached {
```

Comment:
Perheps we don't need that helper at all? It does not save us from much of the boilerplate, but introduces complexity in the form of Kotlin's delegates.